content/blog/purrr-1-0-0/index.Rmd

```diff
@@ -0,0 +1,485 @@
+Previously, these functions had inconsistent behaviour when you attempted to modify an element with `NULL`: some functions would delete that element, and some would set it to `NULL`.
+That inconsistency arose because base R handles `NULL` in different ways depending on whether or not use you `$`/`[[` or `[`:
+
+```{r}
+x1 <- x2 <- x3 <- list(a = 1, b = 2)
+
+x1$a <- NULL
+str(x1)
+
+x2["a"] <- list(NULL)
+str(x2)
+```
+
+Now all list modifying functions will create an element containing `NULL`:
+
+```{r}
+x3 |> 
+  list_modify(a = NULL) |> 
+  str()
+
+x3 |> 
+  modify_at("b", \(x) NULL) |> 
+  str()
+```
+
+If you want to delete the element, you can use the special `zap()` sentinel:
+
+```{r}
+x3 |> 
+  list_modify(a = zap()) |> 
+  str()
+```
+
+`zap()` does not work in `modify*()` because those functions are designed to always return the same top-level structure as the input.
+
+### Core purpose refinements
+
+We have **deprecated** a number of functions to keep purrr focused on its core purpose: facilitating functional programming in R.
+Deprecation means that the functions will continue to work, but you'll be warned once every 8 hours if you use them.
+In several years time, we'll release an update which causes the warnings to occur on every time you use them, and a few years after that they'll transformed to throwing errors.
+
+-   `cross()` and all its variants have been deprecated because they're slow and buggy, and a better approach already exists in `tidyr::expand_grid()`.
+
+-   `update_list()`, `rerun()`, and the use of tidyselect with `map_at()` and friends have been deprecated because we no longer believe that non-standard evaluation is a good fit for purrr.
+
+-   The `lift_*` family of functions has been deprecated because they promote a style of function manipulation that is not commonly used in R.
+
+-   `prepend()`, `rdunif()`, `rbernoulli()`, `when()`, and `list_along()` have been deprecated because they're not directly related to functional programming.
+
+-   `splice()` has been deprecated because we no longer believe that automatic splicing makes for good UI and there are other ways to achieve the same result.
+
+Consult the documentation for the alternatives that we now recommend.
+
+Deprecating these functions makes purrr easier to maintain because it reduces the surface area for bugs and issues, and it makes purrr easier to learn because there's a clearer common thread that ties together all functions.
+
+## Documentation
+
+As you've seen in the code above, we are moving from magrittr's pipe (`%>%`) to the base pipe (`|>`) and from formula syntax (`~ .x + 1`) to R's new anonymous function short hand (`\(x) x + 1`).
+We believe that it's better to use these new base tools because they work everywhere: the base pipe doesn't requite that you load magrittr and the new function shorthand works everywhere, not just in purrr functions.
```

Comment:
```suggestion
We believe that it's better to use these new base tools because they work everywhere: the base pipe doesn't require that you load magrittr and the new function shorthand works everywhere, not just in purrr functions.
```